File: src/ffi_esgleam.mjs
// @ts-check
import { spawnSync } from 'node:child_process';
import { chmod, mkdir, writeFile } from 'node:fs/promises';
import { watch } from 'node:fs';
import { resolve } from 'node:path';
import { Buffer } from 'node:buffer';
import { default as process } from 'node:process';
import {
  Win32,
  Linux,
  Darwin,
  Solaris,
  Freebsd,
  Openbsd,
  Arm,
  Arm64,
  Ia32,
  Ppc64,
  X64,
  // @ts-expect-error
} from './esgleam/mod/platform.mjs';
// @ts-expect-error
import { Ok, Error } from './gleam.mjs';

import { entries } from './streaming_tar.mjs';

const install_dir = "./build/dev/bin"
const pkg_dir = `${install_dir}/package/bin`;

export function exec_shell(command) {
  command = command.toArray();
  spawnSync(command[0], command.slice(1), { cwd: '.', stdio: 'inherit' });
}

/** @type {Partial<Record<NodeJS.Platform, () => import('../build/dev/javascript/esgleam/esgleam/mod/platform.d.mts').OsName$>>} */
const platform_map = {
  darwin: () => new Darwin(),
  freebsd: () => new Freebsd(),
  linux: () => new Linux(),
  openbsd: () => new Openbsd(),
  sunos: () => new Solaris(),
  win32: () => new Win32(),
};

export function get_os() {
  const platform = process.platform;
  const res = platform_map[platform]?.();
  if (res) return new Ok(res);
  return new Error(res);
}

/** @type {Partial<Record<NodeJS.Architecture, () => import('../build/dev/javascript/esgleam/esgleam/mod/platform.d.mts').Arch$>>} */
const arch_map = {
  arm: () => new Arm(),
  arm64: () => new Arm64(),
  ia32: () => new Ia32(),
  ppc64: () => new Ppc64(),
  x64: () => new X64(),
};

export function get_arch() {
  const arch = process.arch;
  const res = arch_map[arch]?.();
  if (res) return new Ok(res);
  return new Error(res);
}

/**
 * @param {string} tarball_url 
 * @param {VoidFunction} then 
 */
export async function do_fetch(tarball_url, then) {
  console.log(`Fetching tarball from: ${tarball_url}`);
  const tarResp = await fetch(tarball_url);
  const tarStream = tarResp.body?.pipeThrough(new DecompressionStream('gzip'));

  for await (const entry of entries(tarStream)) {
    if (/(^|\/)esbuild(\.exe)?$/.test(entry.name)) {
      try {
        await mkdir(pkg_dir, { recursive: true });
      } catch {}
      const exePath = entry.name.endsWith('.exe') ? `${pkg_dir}/esbuild.exe` : `${pkg_dir}/esbuild`;
      await writeFile(
        exePath,
        Buffer.from(await entry.arrayBuffer()),
        { encoding: 'binary' }
      );
      await chmod(exePath, 0o777);
      break;
    }
  }
  then()
}

export function install() {
  spawnSync('gleam', ['run', '--target=erlang', '-m', 'esgleam/install'], {stdio: 'inherit'});
}

export function watch_gleam() {
  const _path = resolve('./src');
  console.log(_path, process.cwd());
  /** @type {import("node:fs").FSWatcher} */
  const _watcher = watch(_path, { recursive: true }, (event, filename) => {
    console.log(`[${event}]: ${filename}`);
    spawnSync('gleam', ['build', '--target=javascript'], {stdio: 'inherit'});
  });
  return () => {
    _watcher.close();
  };
}

File: src/ffi_esgleam.erl
-module(ffi_esgleam).

-export([get_arch/0, get_os/0, do_fetch/1, do_exec/2]).

install_dir() -> "./build/dev/bin".
% pkg_dir() -> install_dir() + "/package/bin".

-spec get_arch() -> {ok, x64 | ia32 | arm64 | arm | ppc64} | {err, string()}.
get_arch() ->
    % #TODO: find out how to detect if it's not ia32/x64 for windows :P
    case erlang:system_info(os_type) of
        {unix, _} -> 
            [Arch_Raw, _] = string:split(erlang:system_info(system_architecture), "-"),
            % based on the values that `uname -a` returns
            case Arch_Raw of
                "x86_64" -> 
                    {ok, x64};
                "amd64" ->
                    {ok, x64};
                "i386" ->
                    {ok, ia32};
                "i586" ->
                    {ok, ia32};
                "i686" ->
                    {ok, ia32};
                "aarch64" ->
                    {ok, arm64};
                "armv7l" ->
                    {ok, arm};
                "ppc64" ->
                    {ok, ppc64};
                _ -> {err, Arch_Raw}
                end;
        {win32, _} -> 
            case erlang:system_info(wordsize) of
                4 -> {ok, ia32};
                8 -> {ok, x64}
            end
    end.

% list compiled from erlang otp source
-spec get_os() -> {ok, win32 | darwin | linux | solaris | freebsd | openbsd | dragonfly | irix64 | irix | posix} | {err, atom()}.
get_os() ->
    case erlang:system_info(os_type) of
        {win32, _} -> win32;
        {unix, linux} -> {ok, linux};
        {unix, darwin} -> {ok, darwin};
        {unix, sunos} -> {ok, solaris};
        {unix, sunos4} -> {ok, sunos}; % technically not solaris
        {unix, freebsd} -> {ok, freebsd};
        {unix, openbsd} -> {ok, openbsd};
        {unix, Os} -> {err, Os}
    end.

-spec do_fetch(string()) -> {}.
do_fetch(Url) ->
    inets:start(),
    ssl:start(),
    {ok, {{_, 200, _}, _, Body}} = httpc:request(get, {Url, []}, [], []),
    {ok, Res} = thoas:decode(Body),
    #{<<"dist">> := #{<<"tarball">> := TarUrl}} = Res,
    fetch_tarball(TarUrl)
.

-spec fetch_tarball(string()) -> {}.
fetch_tarball(Url) ->
    io:fwrite("Fetching tarball from: ~s\n", [Url]),
    {ok, {{_, 200, _}, _, Binary}} = httpc:request(get, {Url, []}, [], [{body_format, binary}]),
    case get_os() of
        {win32, _} -> {
            ok = erl_tar:extract({binary, Binary}, [compressed, verbose, {cwd, install_dir()}, {files, ["package/bin/esbuild.exe"]}])
        };
        _ -> {
            ok = erl_tar:extract({binary, Binary}, [compressed, verbose, {cwd, install_dir()}, {files, ["package/bin/esbuild"]}])
        }
    end
.

-spec do_exec(string(), list(string())) -> a.
do_exec(Cmd, Args) ->
    % Exec = case get_os() of
    %     {win32, _} -> "./priv/bin/esbuild.exe";
    %     _ -> "./priv/bin/esbuild"
    % end,
    Port = open_port({spawn_executable, Cmd}, [exit_status, stderr_to_stdout, {args, Args}]),
    do_recieve(Port)
    % receive {Port, {exit_status, Code}} -> Code end
    % get_port_data(Port)
.

do_recieve(Port) ->
    receive 
        {Port, {data, Data}} ->
            BinaryData = iolist_to_binary(Data),
            io:fwrite("~ts\n", [BinaryData]),
            do_recieve(Port)
        ;
        {Port, closed} -> {ok};
        {'EXIT', Port, Reason} -> 
            exit(Reason),
            {err, Reason};
        {Port, {exit_status, Code}} ->
            {ok, Code};
        _ -> {err, "Something went wrong"}
    end
.
